specifyweb/frontend/js_src/lib/components/WorkBench/WbUtils.tsx

```diff
@@ -6,35 +6,232 @@
+  readonly utils: WbUtils;
+  readonly spreadsheetContainer: any;
+}): JSX.Element {
+  const [currentPosition, setCurrentPosition] = React.useState<number>(0);
+  const [buttonIsPressed, _press, _unpress, togglePress] = useBooleanState();
+  const handleTypeToggle = () => {
+    togglePress();
+    utils.toggleCellTypes(name, 'toggle', spreadsheetContainer.current);
+  };
+  const handlePrevious = () => {
+    const [_, position] = utils.navigateCells({
+      type: name,
+      direction: 'previous',
+      currentCellPosition: currentPosition,
+      totalCount,
+    });
+    setCurrentPosition(position);
+  };
+  const handleNext = () => {
+    const [_, position] = utils.navigateCells({
+      type: name,
+      direction: 'next',
+      currentCellPosition: currentPosition,
+      totalCount,
+    });
+    setCurrentPosition(position);
+  };
+  return (
+    <span
+      aria-atomic
+      className="wb-navigation-section flex rounded"
+      data-navigation-type={name}
+    >
+      <Button.Small
+        className="brightness-80 hover:brightness-70 p-2 ring-0"
+        data-navigation-direction="previous"
+        variant="bg-inherit text-gray-800 dark:text-gray-100"
+        onClick={handlePrevious}
+      >
+        {icons.chevronLeft}
+      </Button.Small>
+      <Button.Small
+        className={`
+          hover:brightness-70 grid grid-cols-[auto_1fr_auto_1fr_auto]
+          items-center ring-0
+          ${className.ariaHandled}
+          ${buttonIsPressed ? 'brightness-50' : ''}
+        `}
+        aria-pressed={buttonIsPressed}
+        title={wbText.clickToToggle()}
+        variant="bg-inherit text-gray-800 dark:text-gray-100"
+        onClick={handleTypeToggle}
+      >
+        {label} (
+        <span className="text-center">
+          {currentPosition}
+        </span>
+        /<span>{totalCount}</span>)
+      </Button.Small>
+      <Button.Small
+        className="brightness-80 hover:brightness-70 p-2 ring-0"
+        data-navigation-direction="next"
+        type="button"
+        variant="bg-inherit text-gray-800 dark:text-gray-100"
+        onClick={handleNext}
+      >
+        {icons.chevronRight}
+      </Button.Small>
+    </span>
+  );
+}
+
+export function WbUtilsComponent({
+  isUploaded,
+  cellCounts,
+  utils,
+  spreadsheetContainer,
+}: {
+  readonly isUploaded: boolean;
+  readonly cellCounts: WbCellCounts;
+  readonly utils: WbUtils;
+  readonly spreadsheetContainer: any;
+}): JSX.Element {
+  const searchRef = React.useRef<HTMLInputElement | null>(null);
+  const replaceRef = React.useRef<HTMLInputElement | null>(null);
+
+  const handleSearch = (event: KeyboardEvent) => {
+    // TODO: check if debounce is needed here
+    utils.searchCells(
+      event,
+      searchRef.current as HTMLInputElement,
+      spreadsheetContainer.current
+    );
+  };
+
+  const handleReplace = (event: KeyboardEvent) => {
+    utils.replaceCells(event, replaceRef.current as HTMLInputElement);
```

Comment:
```suggestion
    utils.replaceCells(event, replaceRef.current!);
```

generally try to not have no type assertion if possible, and if not (or not convenient), use the weakest type assertion you can (in this case `!` just removes the `| null` part, so is weaker, where as ` as ` can be used to cast to a very different type - which is bad as it can hide bugs on code refactoring